docs/src/main/asciidoc/websockets-next-reference.adoc

```diff
@@ -883,6 +886,83 @@ quarkus.http.auth.permission.http-upgrade.paths=/end
 quarkus.http.auth.permission.http-upgrade.policy=authenticated
 ----
 
+=== Secure server endpoints with permission checkers
+
+WebSocket endpoints can be secured with the xref:security-authorize-web-endpoints-reference.adoc#permission-checker[permission checkers].
+We strongly recommend to revise the <<secure-http-upgrade>> section and secure the HTTP upgrade like in the example below:
+
+[source, java]
+----
+@PermissionsAllowed("product:premium")
+@WebSocket(path = "/product/premium")
+public class PremiumProductEndpoint {
+
+    @OnTextMessage
+    PremiumProduct getPremiumProduct(int productId) {
+        return new PremiumProduct();
+    }
+
+}
+
+@ApplicationScoped
+public class PermissionChecker {
+
+    @PermissionChecker("product:premium")
+    public boolean canGetPremiumProduct(SecurityIdentity securityIdentity) { <1>
+        String username = currentIdentity.getPrincipal().getName();
+        return isPremiumCustomer(username);
+    }
+
+}
+----
+<1> A permission checker authorizing an HTTP upgrade must declare exactly one method parameter, the `SecurityIdentity`.
```

Comment:
Is it too restrictive to require a single parameter, of specific type ? HTTP upgrade can contain query parameters, for example, for the custom ticketing system, https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/secure-sql-chatbot#custom-websocket-ticket-system

Reply:
This is only valid for HTTP upgrade.

I don't mind to support more arguments, but this restriction comes from the fact that it is a `@PermissionsAllowed` placed on a class, so only `SecurityIdentity` is supported in Quarkus Security. The reason for this is that in general, where would you take these other arguments from if it is not applied on a method? From CDI or .... We can implement that OR part. Which do you have in mind? If it is about `RoutingContext`, I presume it will be available on `SecurityIdentity` if it is OIDC or basic for auth.

Reply:
maybe showing how to access routing context in the example would be useful

Reply:
I was also thinking that if user needs something really special, they could secure `@OnOpen` (cc @mkouba ) which would be very similar, just bit more processing but allows you to inject connection. But I was too tired yesterday to create test so I let it be.

Reply:
Hi @michalvavrik Sure, custom ticketing system is not a typical case, let's keep it as is

Reply:
@sberyozkin I was wrong, not sure why I remember it incorrectly (maybe I am not looking right), but I only found OIDC identity provider adds the `RoutingContext`. So I made couple of changes to ensure that `RoutingContext` is always propagated from authentication request (where we know it is as builtin auth mechanisms add it) to the identity on which permission check is performed.

I am not exactly sure how you will feel about changes, I think they are not super pretty but acceptable. if you have strong opinion, let me know and I'll revert it.